README.md
# logger
golang logger
简单日志


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.6](https://github.com/shuqingzai/logger/compare/v0.0.5...v0.0.6) (2021-06-14)


### Features

* **file(文件驱动):** 1. 支持配置 file_map_size 设置文件句柄map的初始化数量,2. 支持配置 log_split_type 设置文件切割方式,3. 支持配置 log_split_size 设置文件切割大小, 3 支持 log_file_ext 设置日志文件后缀 .txt或.log ([7f194f9](https://github.com/shuqingzai/logger/commit/7f194f90c3157ddeb48e99813e4fe8d0697bc432))

### [0.0.5](https://github.com/shuqingzai/logger/compare/v0.0.4...v0.0.5) (2021-06-13)


### Features

* **file(文件驱动):** 支持配置切割文件方式: 1 支持配置按小时切割文件 2 支持配置按文件大小切割文件 ([6a636cd](https://github.com/shuqingzai/logger/commit/6a636cd38e20fbf5a51abbd5670e786f63ae17f8))
* **other(其他):** 文件驱动初始化统一入口 InitLogger() ([bcbd3b4](https://github.com/shuqingzai/logger/commit/bcbd3b4b20c36e95ff5493f88b521656ec37de47))

### [0.0.4](https://github.com/shuqingzai/logger/compare/v0.0.3...v0.0.4) (2021-06-13)

### [0.0.3](https://github.com/shuqingzai/logger/compare/v0.0.2...v0.0.3) (2021-06-13)

### 0.0.2 (2021-06-13)


### Features

* **file(文件驱动):** 新增检查文件后缀，如果不是 .log结尾的文件后缀，自动补上 ([3591562](https://github.com/shuqingzai/logger/commit/3591562cacaa69e99f6a4f38a2d764715ec709f5))
* **other(其他):** 首次commit ([2ede038](https://github.com/shuqingzai/logger/commit/2ede038f9c13dff71050bc00ad43f5ff7184f76c))


logger_test.go
package logger

import (
	"fmt"
	"testing"
	"time"
)

func TestInitLogger(t *testing.T) {
	config := make(map[string]string, 8)
	//config["log_path"] = "logs"
	//config["log_name"] = "test1"
	//config["file_map_size"] = "30" // 30 -300
	//config["log_chan_size"] = "50000" // 1 - 10000
	//config["log_split_type"] = "size" // size || hour
	//config["log_split_size"] = "1" // 单位: MB
	//config["log_file_ext"] = ".log" // .txt || .log
	logger, err := InitLogger("file", config)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer logger.Close()
	for {
		logger.Debug("testDebug %d", time.Now().Unix())
		logger.Trace("testTrace %d", time.Now().Unix())
		logger.Info("testInfo %d", time.Now().Unix())
		logger.Warn("testWarn %d", time.Now().Unix())
		logger.Error("testError %d", time.Now().Unix())
		logger.Fatal("testFatal %d", time.Now().Unix())
		time.Sleep(time.Second)
	}
}

func TestFileLogger(t *testing.T) {
	config := make(map[string]string, 8)
	logger, err := NewFileLogger(config)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer logger.Close()
	logger.Debug("testDebug %d", time.Now().Unix())
	logger.Trace("testTrace %d", time.Now().Unix())
	logger.Info("testInfo %d", time.Now().Unix())
	logger.Warn("testWarn %d", time.Now().Unix())
	logger.Error("testError %d", time.Now().Unix())
	logger.Fatal("testFatal %d", time.Now().Unix())
}

func TestConsoleLogger(t *testing.T) {

	config := map[string]string{}
	logger, err := NewConsoleLogger(config)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer logger.Close()
	logger.Debug("testDebug %d", time.Now().Unix())
	logger.Trace("testTrace %d", time.Now().Unix())
	logger.Info("testInfo %d", time.Now().Unix())
	logger.Warn("testWarn %d", time.Now().Unix())
	logger.Error("testError %d", time.Now().Unix())
	logger.Fatal("testFatal %d", time.Now().Unix())
}


util.go
package logger

import (
	"fmt"
	"os"
	"runtime"
	"strings"
	"time"
)

// getLineInfo
// 获取函数调用栈信息
// @return fileName
// @return funcName
// @return lineNo
// @Author ShuQingZai<overbeck.jack@qq.com>
func getLineInfo(skip int) (fileName string, funcName string, lineNo int) {
	pc, file, line, ok := runtime.Caller(skip)
	if ok {
		fileName = file
		funcName = runtime.FuncForPC(pc).Name()
		lineNo = line
	}
	return
}

// writeFile
// 写入文件
// @param fileHandle
// @param levelStr
// @param format
// @param args
// @Author ShuQingZai<overbeck.jack@qq.com>
func writeFile(skip int, level int, format string, args ...interface{}) (logData *LogData) {
	// 文件写入
	nowTime := time.Now().Format("2006-01-02 15:04:05.999")
	fileName, funcName, lineNo := getLineInfo(skip)
	msg := fmt.Sprintf(format, args...)
	logData = &LogData{
		LogTime:  nowTime,
		Level:    level,
		Message:  msg,
		FileName: fileName,
		LineNo:   lineNo,
		FuncName: funcName,
	}

	return
}

func checkDir(path string) error {
	dirPath, err := os.Stat(path)
	if err == nil {
		if dirPath.IsDir() {
			return nil
		}
		return fmt.Errorf("%s is Exist, but it is not a dir", path)
	}
	if os.IsNotExist(err) {
		return os.Mkdir(path, 0755)
	}

	return fmt.Errorf("checkDir error: %s", err)
}

func checkFileName(fileName string, fileSuffix string) string {
	if strings.HasSuffix(fileName, fileSuffix) {
		return fileName
	}

	return fmt.Sprintf("%s%s", fileName, fileSuffix)
}


.cz-config.js
module.exports = {
    types: [
        {value: 'feat', name: 'feat:     一个新的特性'},
        {value: 'fix', name: 'fix:      修复一个Bug'},
        {value: 'docs', name: 'docs:     更新的只是文档'},
        {
            value: 'style',
            name:
                'style:    空格, 分号等格式修复',
        },
        {
            value: 'refactor',
            name: 'refactor: 代码重构，注意和特性、修复区分开',
        },
        {
            value: 'perf',
            name: 'perf:     提升性能',
        },
        {value: 'test', name: 'test:     添加一个测试'},
        {
            value: 'chore',
            name:
                'chore:    开发工具变动(构建、脚手架工具等)',
        },
        {value: 'revert', name: 'revert:   代码回退'},
        {value: 'WIP', name: 'WIP:      代码迭代'},
    ],

    scopes: [
        {name: 'file(文件驱动)'},
        {name: 'console(控制台驱动)'},
        {name: 'sql(数据库)'},
        {name: 'other(其他)'},
    ],

    allowTicketNumber: false,
    isTicketNumberRequired: false,
    ticketNumberPrefix: 'TICKET-',
    ticketNumberRegExp: '\\d{1,5}',

    // it needs to match the value for field type. Eg.: 'fix'
    /*
    scopeOverrides: {
      fix: [
        {name: 'merge'},
        {name: 'style'},
        {name: 'e2eTest'},
        {name: 'unitTest'}
      ]
    },
    */
    // override the messages, defaults are as follows
    messages: {
        type: "选择一种你的提交类型:",
        scope: '\n选择一个scope (可选):',
        // used if allowCustomScopes is true
        customScope: '表示此次更改的范围:',
        subject: '简短短说明:\n',
        body: '详细说明，使用"|"换行(可选):\n',
        breaking: '非兼容性说明 (可选):\n',
        footer: '关联关闭的issue，例如：#31, #34(可选):\n',
        confirmCommit: '确定提交说明?',
    },

    allowCustomScopes: true,
    allowBreakingChanges: ['feat', 'fix'],
    // skip any questions you want
    skipQuestions: ['body'],

    // limit subject length
    subjectLimit: 100,
    // breaklineChar: '|', // It is supported for fields body and footer.
    footerPrefix: 'ISSUES CLOSED:'
    // askForBreakingChangeFirst : true, // default is false
};

consoleLogger.go
package logger

import (
	"fmt"
	"os"
)

type ConsoleLogger struct {
}

func (c *ConsoleLogger) Debug(format string, args ...interface{}) {
	c.writeLog(LogLevelDebug, format, args...)
}

func (c *ConsoleLogger) Trace(format string, args ...interface{}) {
	c.writeLog(LogLevelTrace, format, args...)
}

func (c *ConsoleLogger) Info(format string, args ...interface{}) {
	c.writeLog(LogLevelInfo, format, args...)
}

func (c *ConsoleLogger) Warn(format string, args ...interface{}) {
	c.writeLog(LogLevelWarn, format, args...)
}

func (c *ConsoleLogger) Error(format string, args ...interface{}) {
	c.writeLog(LogLevelError, format, args...)
}

func (c *ConsoleLogger) Fatal(format string, args ...interface{}) {
	c.writeLog(LogLevelFatal, format, args...)
}

func (c *ConsoleLogger) Close() {

}

func NewConsoleLogger(config map[string]string) (c LogInterface, err error) {
	c = &ConsoleLogger{}
	return
}

func (c *ConsoleLogger) writeLog(level int, format string, args ...interface{}) {
	logData := writeFile(4, level, format, args...)
	_, err := fmt.Fprintf(os.Stdout, "%s %s [%s::%d %s()]\n%s\n",
		logData.LogTime,
		getLogLevelText(logData.Level),
		logData.FileName,
		logData.LineNo,
		logData.FuncName,
		logData.Message)
	if err != nil {
		panic(fmt.Sprintf("Write Log error: %s", err))
	}
}


fileLogger.go
package logger

import (
	"fmt"
	"os"
	"path"
	"strconv"
	"strings"
	"time"
)

type FileLogger struct {
	level         int
	logFilePath   string
	logFilename   string
	logFileExt    string
	splitType     int
	splitSize     int64
	lastSplitHour int
	fileMapSize   int
	logChanSize   int
	fileMap       map[string]*os.File
	logDataChan   chan *LogData
}

func (f *FileLogger) Close() {
	for k, file := range f.fileMap {
		if file != nil {
			_ = file.Close()
		}
		delete(f.fileMap, k)
	}
}

func (f *FileLogger) Debug(format string, args ...interface{}) {
	f.writeLog(LogLevelDebug, format, args...)
}

func (f *FileLogger) Trace(format string, args ...interface{}) {
	f.writeLog(LogLevelTrace, format, args...)
}

func (f *FileLogger) Info(format string, args ...interface{}) {
	f.writeLog(LogLevelInfo, format, args...)
}

func (f *FileLogger) Warn(format string, args ...interface{}) {
	f.writeLog(LogLevelWarn, format, args...)
}

func (f *FileLogger) Error(format string, args ...interface{}) {
	f.writeLog(LogLevelError, format, args...)
}

func (f *FileLogger) Fatal(format string, args ...interface{}) {
	f.writeLog(LogLevelFatal, format, args...)
}

func NewFileLogger(config map[string]string) (LogInterface, error) {
	logPath, ok := config["log_path"]
	if !ok {
		logPath = "logs"
	}

	err := checkDir(logPath)
	if err != nil {
		return nil, err
	}

	logFilename, ok := config["log_name"]
	if !ok {
		logFilename = "logger"
	}

	logChanSizeStr, ok := config["log_chan_size"]
	if !ok {
		logChanSizeStr = "50000"
	}
	logChanSize, err := strconv.Atoi(logChanSizeStr)
	if err != nil || logChanSize < 1 || logChanSize > 100000 {
		logChanSize = 50000
	}
	logFileExt := config["log_file_ext"]
	if logFileExt != ".log" && logFileExt != ".txt" {
		logFileExt = ".log"
	}

	fileMapSizeStr, ok := config["file_map_size"]
	if !ok {
		fileMapSizeStr = "30"
	}
	fileMapSize, err := strconv.Atoi(fileMapSizeStr)
	if err != nil || fileMapSize < 30 || fileMapSize > 300 {
		fileMapSize = 30
	}

	f := &FileLogger{
		level:         LogLevelDebug,
		logFilePath:   logPath,
		logFilename:   checkFileName(logFilename, logFileExt),
		logFileExt:    logFileExt,
		fileMapSize:   fileMapSize,
		fileMap:       make(map[string]*os.File, fileMapSize),
		logChanSize:   logChanSize,
		logDataChan:   make(chan *LogData, logChanSize),
		lastSplitHour: time.Now().Hour(),
	}
	// 日志切割方式
	logSplitTypeStr, ok := config["log_split_type"]
	if !ok {
		logSplitTypeStr = "size"
	}
	if logSplitTypeStr == "size" {
		f.splitType = LogSplitTypeSize
		logSplitSizeStr, ok := config["log_split_size"]
		if !ok {
			logSplitSizeStr = "20"
		}
		logSplitSize, err := strconv.ParseInt(logSplitSizeStr, 10, 64)
		if err != nil {
			logSplitSize = 20
		}
		f.splitSize = logSplitSize * 1024 * 1024 // 单位是Mb
	} else {
		f.splitType = LogSplitTypeHour
	}

	// 启动协程写入日志文件
	go f.saveLogFile()

	return f, err
}

func (f *FileLogger) backupFile(logFilePath string, fileHandle *os.File, fileMapKey string, now *time.Time) (err error) {
	err = fileHandle.Close()
	if err != nil {
		return
	}
	delete(f.fileMap, fileMapKey) // 删除Map元素

	backupFilename := fmt.Sprintf("%s_%04d_%02d_%02d_%02d_%02d_%02d%s",
		strings.Replace(logFilePath, f.logFileExt, "", -1),
		now.Year(),
		now.Month(),
		now.Day(),
		f.lastSplitHour,
		now.Minute(),
		now.Second(),
		f.logFileExt,
	)
	f.lastSplitHour = now.Hour()
	err = os.Rename(logFilePath, backupFilename)

	return
}

func (f FileLogger) splitFileOfHour(logFilePath string) {
	now := time.Now()
	if now.Hour() == f.lastSplitHour {
		return
	}
	fileHandle, ok, fileMapKey := f.checkFileHandle(logFilePath)
	if !ok {
		return
	}
	_ = f.backupFile(logFilePath, fileHandle, fileMapKey, &now)
}

func (f *FileLogger) splitFileOfSize(logFilePath string) {
	fileHandle, ok, fileMapKey := f.checkFileHandle(logFilePath)
	if !ok {
		return
	}
	statInfo, err := fileHandle.Stat()
	if err != nil {
		return
	}
	if statInfo.Size() < f.splitSize {
		return
	}
	now := time.Now()
	_ = f.backupFile(logFilePath, fileHandle, fileMapKey, &now)
}

func (f *FileLogger) checkSplitFile(logFilePath string) {
	if f.splitType == LogSplitTypeHour {
		// 按小时切割文件
		f.splitFileOfHour(logFilePath)
		return

	}

	// 按文件大小切割文件
	f.splitFileOfSize(logFilePath)
}

func (f *FileLogger) saveLogFile() {
	for logData := range f.logDataChan {
		// 检查文件是否需要切割
		f.checkSplitFile(logData.LogFilePath)
		fileHandle := f.getFileHandle(logData.LogFilePath)
		_, _ = fmt.Fprintf(fileHandle, "%s %s [%s::%d %s()]\n%s\n",
			logData.LogTime,
			logData.LevelStr,
			logData.FileName,
			logData.LineNo,
			logData.FuncName,
			logData.Message)
	}
}

func (f *FileLogger) openFile(filePath string) *os.File {
	file, err := os.OpenFile(filePath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		panic(fmt.Sprintf("Open file %s failed, err: %s ", filePath, err))
	}

	return file
}

func (f *FileLogger) getFileHandle(filePath string) *os.File {
	fileHandle, ok, fileMapKey := f.checkFileHandle(filePath)
	if !ok {
		// 打开文件并存储在fileMap
		fileHandle = f.openFile(filePath)
		f.fileMap[fileMapKey] = fileHandle
	}

	return fileHandle
}

func (f *FileLogger) checkFileHandle(filePath string) (fileHandle *os.File, ok bool, fileMapKey string) {
	fileMapKey = path.Base(filePath)
	fileHandle, ok = f.fileMap[fileMapKey]
	return
}

func (f *FileLogger) getFilePath(fileName string) (filePath string, levelStr string) {
	dateTime := time.Now().Format("2006-01-02")
	levelStr = getLogLevelText(f.level)
	filePath = fmt.Sprintf("%s/%s_%s_%s", f.logFilePath, dateTime, levelStr, fileName)
	return
}

func (f *FileLogger) setLogLevel(level int) {
	if level >= LogLevelDebug && level <= LogLevelFatal {
		f.level = level
	}
}

func (f *FileLogger) writeLog(level int, format string, args ...interface{}) {
	f.setLogLevel(level)
	// 解析获取文件路径与错误级别描述
	filePath, levelStr := f.getFilePath(f.logFilename)
	// 获取文件句柄并存储在fileMap中
	//fileHandle := f.getFileHandle(filePath)
	// 组装信息放入chan队列
	logData := writeFile(4, level, format, args...)
	logData.LogFilePath = filePath
	logData.LevelStr = levelStr
	// 放入chan 如果chan已满，直接跳过
	select {
	case f.logDataChan <- logData:
	default:
	}
}
